packages/credential-w3c/src/action-handler.ts

```diff
@@ -573,11 +612,17 @@ export class CredentialIssuer implements IAgentPlugin {
           challenge: args.challenge,
           domain: args.domain,
           audience,
+          policies: args.policies
         })
-        return true
+        return { verified: true }
       } catch (e: any) {
-        //TODO: return a more detailed reason for failure
-        return false
+        const errorCodes = ['invalid_jwt','invalid_config', 'invalid_signature', 'not_supported', 'no_suitable_keys', 'resolver_error']
+        if (errorCodes.some(errorCode => e.message.includes(errorCode))) {
+          return { verified: false, details: e.message }
+        }
+        else {
+          throw e
```

Comment:
I really like this separation of concerns between legitimate errors and exceptional situations.

However, I'm not yet 100% sure we should be introducing an extra branch, though.
I'm concerned about the usability here where developers have to handle 3 situations: `valid response`, `invalid response` and `thrown error`.
This is not a critique of your approach (which I've also described in the past), but rather me venting my uncertainty :)

Reply:
Understandable, how do you propose we approach this issue?

Reply:
I suppose the simple answer would be to not re-throw an error in the exceptional case, but to return an object just like for the legitimate failures.
The difference would be that for the legitimate failures we can also identify the cause and populate the  `errorCode` property of the error.